ext/posix/posix.c

```diff
@@ -426,11 +426,10 @@ static int php_posix_stream_get_fd(zval *zfp, zend_long *fd) /* {{{ */
 	if (stream == NULL) {
 		return 0;
 	}
-	if (php_stream_can_cast(stream, PHP_STREAM_AS_FD_FOR_SELECT) == SUCCESS) {
-		php_stream_cast(stream, PHP_STREAM_AS_FD_FOR_SELECT, (void*)fd, 0);
-	} else if (php_stream_can_cast(stream, PHP_STREAM_AS_FD) == SUCCESS) {
-		php_stream_cast(stream, PHP_STREAM_AS_FD, (void*)fd, 0);
-	} else {
+	if (
+		php_stream_cast(stream, PHP_STREAM_AS_FD_FOR_SELECT, (void*)&fd, 0) == FAILURE &&
+		php_stream_cast(stream, PHP_STREAM_AS_FD, (void*)&fd, 0) == FAILURE
+	) {
```

Comment:
We should be cautious about this, because `php_stream_cast` can have side effects even if it fails. For example, `php_stream_cast(stream, PHP_STREAM_AS_FD)` may fflush + fseek the stream, and return FAILURE later.

This could be a breaking change, as posix_isatty() calls that used to fail without have a side effect the stream may now have a side effect.

Reply:
Oh, indeed! I focused on `show_err` only, but depending on whether `ret` is set, the logic is certainly different.

Reply:
@arnaud-lb Can you give an example of such side effects in the code? Just wondering if I missed something specific or this is just a theoretical concern about API and possibly some 3rd party stream wrapper?

Reply:
I have done some checking and don't see any case where FAILURE would be returned for can_cast and SUCCESS for cast.

Reply:
> Can you give an example of such side effects in the code? 

php_stream_flush and ops->seek() here are called only in cast, and not in can_cast:

https://github.com/php/php-src/blob/5f42a46405c780e79d17c4bdf0f9f0e244a8b9d1/main/streams/cast.c#L197-L202

In https://gist.github.com/arnaud-lb/700842ba8c72b9b3e44ff3c5cc7763f4, stream_seek() is called after this change

Reply:
Ah you are right of course. Completely missed that bit as I was just checking the actual cast ops usage.